src/zenml/integrations/kubernetes/step_operators/kubernetes_step_operator.py

```diff
@@ -0,0 +1,235 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Kubernetes step operator implementation."""
+
+from typing import TYPE_CHECKING, Dict, List, Optional, Tuple, Type, cast
+
+from kubernetes import client as k8s_client
+
+from zenml.config.base_settings import BaseSettings
+from zenml.config.build_configuration import BuildConfiguration
+from zenml.enums import StackComponentType
+from zenml.integrations.kubernetes.flavors import (
+    KubernetesStepOperatorConfig,
+    KubernetesStepOperatorSettings,
+)
+from zenml.integrations.kubernetes.orchestrators import kube_utils
+from zenml.integrations.kubernetes.orchestrators.manifest_utils import (
+    build_pod_manifest,
+)
+from zenml.logger import get_logger
+from zenml.stack import Stack, StackValidator
+from zenml.step_operators import BaseStepOperator
+
+if TYPE_CHECKING:
+    from zenml.config.base_settings import BaseSettings
+    from zenml.config.step_run_info import StepRunInfo
+    from zenml.models import PipelineDeploymentBase
+
+logger = get_logger(__name__)
+
+KUBERNETES_STEP_OPERATOR_DOCKER_IMAGE_KEY = "kubernetes_step_operator"
+
+
+class KubernetesStepOperator(BaseStepOperator):
+    """Step operator to run on Kubernetes."""
+
+    _k8s_client: Optional[k8s_client.ApiClient] = None
+
+    @property
+    def config(self) -> KubernetesStepOperatorConfig:
+        """Returns the `KubernetesStepOperatorConfig` config.
+
+        Returns:
+            The configuration.
+        """
+        return cast(KubernetesStepOperatorConfig, self._config)
+
+    @property
+    def settings_class(self) -> Optional[Type["BaseSettings"]]:
+        """Settings class for the Kubernetes step operator.
+
+        Returns:
+            The settings class.
+        """
+        return KubernetesStepOperatorSettings
+
+    @property
+    def validator(self) -> Optional[StackValidator]:
+        """Validates the stack.
+
+        Returns:
+            A validator that checks that the stack contains a remote container
+            registry and a remote artifact store.
+        """
+
+        def _validate_remote_components(stack: "Stack") -> Tuple[bool, str]:
+            if stack.artifact_store.config.is_local:
+                return False, (
+                    "The Kubernetes step operator runs code remotely and "
+                    "needs to write files into the artifact store, but the "
+                    f"artifact store `{stack.artifact_store.name}` of the "
+                    "active stack is local. Please ensure that your stack "
+                    "contains a remote artifact store when using the Vertex "
+                    "step operator."
+                )
+
+            container_registry = stack.container_registry
+            assert container_registry is not None
+
+            if container_registry.config.is_local:
```

Comment:
Should we already stop enforcing this given our discussion?

Reply:
I'd make that all part of a separate PR. As long as we don't have the "upload code to artifact store", the container registry is actually mandatory as it won't work with just prebuilt images.